src/solvers/refinement/string_constraint_generator_format.cpp

```diff
@@ -0,0 +1,424 @@
+
+  return true;
+}
+#endif
+
+/// Helper function for parsing format strings.
+/// \param m: a match in a regular expression
+/// \return Format specifier represented by the matched string. The groups in
+///   the match should represent: index, flag, width, precision, date and
+///   conversion type.
+static string_constraint_generatort::format_specifiert
+  format_specifier_of_match(std::smatch &m)
+{
+  string_constraint_generatort::format_specifiert f;
+
+  if(!m[1].str().empty())
+    f.index=std::stoi(m[1].str());
+
+  if(!m[2].str().empty())
+    f.flag=m[2].str();
+
+  if(!m[3].str().empty())
+    f.width=std::stoi(m[3].str());
+
+  if(!m[4].str().empty())
+    f.precision=std::stoi(m[4].str());
+
+  std::string tT=m[5].str();
+  f.dt=(tT!="");
+  if(tT=="T")
+    f.flag.push_back(
+      string_constraint_generatort::format_specifiert::DATE_TIME_UPPER);
+
+  if(m[6].str().length()!=1)
+    throw "invalid format specifier:"+m.str();
+
+  f.conversion=m[6].str()[0];
+  return f;
+}
+
+/// Parse the given string into a format specifiers and text.
+/// \param s: a string
+/// \return A vector of format_elementt.
+static std::vector<format_elementt> parse_format_string(std::string s)
+{
+  std::string format_specifier=
+      "%(\\d+\\$)?([-#+ 0,(\\<]*)?(\\d+)?(\\.\\d+)?([tT])?([a-zA-Z%])";
+  std::regex regex(format_specifier);
+  std::vector<format_elementt> al;
+  std::smatch match;
+
+  while(std::regex_search(s, match, regex))
+  {
+    if(match.position()!= 0)
+    {
+      std::string pre_match=s.substr(0, match.position());
+      al.emplace_back(pre_match);
+    }
+
+    al.emplace_back(format_specifier_of_match(match));
+    s=match.suffix();
+  }
+
+  al.emplace_back(s);
+  return al;
+}
+
+/// Helper for add_axioms_for_format_specifier
+/// \param expr: a structured expression
+/// \param component_name: name of the desired component
+/// \return Expression in the component of `expr` named `component_name`.
+static exprt get_component_in_struct(
+  const struct_exprt &expr, irep_idt component_name)
+{
+  const struct_typet &type=to_struct_type(expr.type());
+  std::size_t number=type.component_number(component_name);
+  return expr.operands()[number];
+}
+
+/// Parse `s` and add axioms ensuring the output corresponds to the output of
+/// String.format. Assumes the argument is a structured expression in which
+/// contains the fields: string expr, int, float, char, boolean, hashcode,
+/// date_time. The correct component will be fetched depending on the format
+/// specifier.
+/// \param fs: a format specifier
+/// \param arg: a struct containing the possible value of the argument to format
+/// \param ref_type: a type  for refined string type
+/// \return String expression representing the output of String.format.
+string_exprt string_constraint_generatort::add_axioms_for_format_specifier(
+  const format_specifiert &fs,
+  const struct_exprt &arg,
+  const refined_string_typet &ref_type)
+{
+  switch(fs.conversion)
```

Comment:
For the unimplemented conversions, you should not return a "close enough" conversion, some should return non-deterministic strings in order to over-approximate the possible behaviour.